// symbol-openapi-typescript-node-client/dist/model/accountRestrictionFlagsEnum.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var AccountRestrictionFlagsEnum;
(function (AccountRestrictionFlagsEnum) {
    AccountRestrictionFlagsEnum[AccountRestrictionFlagsEnum["NUMBER_1"] = 1] = "NUMBER_1";
    AccountRestrictionFlagsEnum[AccountRestrictionFlagsEnum["NUMBER_2"] = 2] = "NUMBER_2";
    AccountRestrictionFlagsEnum[AccountRestrictionFlagsEnum["NUMBER_16385"] = 16385] = "NUMBER_16385";
    AccountRestrictionFlagsEnum[AccountRestrictionFlagsEnum["NUMBER_16388"] = 16388] = "NUMBER_16388";
    AccountRestrictionFlagsEnum[AccountRestrictionFlagsEnum["NUMBER_32769"] = 32769] = "NUMBER_32769";
    AccountRestrictionFlagsEnum[AccountRestrictionFlagsEnum["NUMBER_32770"] = 32770] = "NUMBER_32770";
    AccountRestrictionFlagsEnum[AccountRestrictionFlagsEnum["NUMBER_49153"] = 49153] = "NUMBER_49153";
    AccountRestrictionFlagsEnum[AccountRestrictionFlagsEnum["NUMBER_49156"] = 49156] = "NUMBER_49156";
})(AccountRestrictionFlagsEnum = exports.AccountRestrictionFlagsEnum || (exports.AccountRestrictionFlagsEnum = {}));
