// symbol-openapi-typescript-node-client/dist/model/receiptTypeEnum.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ReceiptTypeEnum;
(function (ReceiptTypeEnum) {
    ReceiptTypeEnum[ReceiptTypeEnum["NUMBER_4685"] = 4685] = "NUMBER_4685";
    ReceiptTypeEnum[ReceiptTypeEnum["NUMBER_4942"] = 4942] = "NUMBER_4942";
    ReceiptTypeEnum[ReceiptTypeEnum["NUMBER_8515"] = 8515] = "NUMBER_8515";
    ReceiptTypeEnum[ReceiptTypeEnum["NUMBER_8776"] = 8776] = "NUMBER_8776";
    ReceiptTypeEnum[ReceiptTypeEnum["NUMBER_9032"] = 9032] = "NUMBER_9032";
    ReceiptTypeEnum[ReceiptTypeEnum["NUMBER_8786"] = 8786] = "NUMBER_8786";
    ReceiptTypeEnum[ReceiptTypeEnum["NUMBER_9042"] = 9042] = "NUMBER_9042";
    ReceiptTypeEnum[ReceiptTypeEnum["NUMBER_12616"] = 12616] = "NUMBER_12616";
    ReceiptTypeEnum[ReceiptTypeEnum["NUMBER_12626"] = 12626] = "NUMBER_12626";
    ReceiptTypeEnum[ReceiptTypeEnum["NUMBER_16717"] = 16717] = "NUMBER_16717";
    ReceiptTypeEnum[ReceiptTypeEnum["NUMBER_16718"] = 16718] = "NUMBER_16718";
    ReceiptTypeEnum[ReceiptTypeEnum["NUMBER_16974"] = 16974] = "NUMBER_16974";
    ReceiptTypeEnum[ReceiptTypeEnum["NUMBER_20803"] = 20803] = "NUMBER_20803";
    ReceiptTypeEnum[ReceiptTypeEnum["NUMBER_57667"] = 57667] = "NUMBER_57667";
    ReceiptTypeEnum[ReceiptTypeEnum["NUMBER_61763"] = 61763] = "NUMBER_61763";
    ReceiptTypeEnum[ReceiptTypeEnum["NUMBER_62019"] = 62019] = "NUMBER_62019";
})(ReceiptTypeEnum = exports.ReceiptTypeEnum || (exports.ReceiptTypeEnum = {}));
