// catbuffer-typescript/builders/AccountAddressRestrictionTransactionBuilder.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var AccountAddressRestrictionTransactionBodyBuilder_1 = require("./AccountAddressRestrictionTransactionBodyBuilder");
var GeneratorUtils_1 = require("./GeneratorUtils");
var TransactionBuilder_1 = require("./TransactionBuilder");
var AccountAddressRestrictionTransactionBuilder = (function (_super) {
    __extends(AccountAddressRestrictionTransactionBuilder, _super);
    function AccountAddressRestrictionTransactionBuilder(signature, signerPublicKey, version, network, type, fee, deadline, restrictionFlags, restrictionAdditions, restrictionDeletions) {
        var _this = _super.call(this, signature, signerPublicKey, version, network, type, fee, deadline) || this;
        _this.accountAddressRestrictionTransactionBody = new AccountAddressRestrictionTransactionBodyBuilder_1.AccountAddressRestrictionTransactionBodyBuilder(restrictionFlags, restrictionAdditions, restrictionDeletions);
        return _this;
    }
    AccountAddressRestrictionTransactionBuilder.loadFromBinary = function (payload) {
        var byteArray = Array.from(payload);
        var superObject = TransactionBuilder_1.TransactionBuilder.loadFromBinary(Uint8Array.from(byteArray));
        byteArray.splice(0, superObject.getSize());
        var accountAddressRestrictionTransactionBody = AccountAddressRestrictionTransactionBodyBuilder_1.AccountAddressRestrictionTransactionBodyBuilder.loadFromBinary(Uint8Array.from(byteArray));
        byteArray.splice(0, accountAddressRestrictionTransactionBody.getSize());
        return new AccountAddressRestrictionTransactionBuilder(superObject.signature, superObject.signerPublicKey, superObject.version, superObject.network, superObject.type, superObject.fee, superObject.deadline, accountAddressRestrictionTransactionBody.restrictionFlags, accountAddressRestrictionTransactionBody.restrictionAdditions, accountAddressRestrictionTransactionBody.restrictionDeletions);
    };
    AccountAddressRestrictionTransactionBuilder.prototype.getRestrictionFlags = function () {
        return this.accountAddressRestrictionTransactionBody.getRestrictionFlags();
    };
    AccountAddressRestrictionTransactionBuilder.prototype.getAccountRestrictionTransactionBody_Reserved1 = function () {
        return this.accountAddressRestrictionTransactionBody.getAccountRestrictionTransactionBody_Reserved1();
    };
    AccountAddressRestrictionTransactionBuilder.prototype.getRestrictionAdditions = function () {
        return this.accountAddressRestrictionTransactionBody.getRestrictionAdditions();
    };
    AccountAddressRestrictionTransactionBuilder.prototype.getRestrictionDeletions = function () {
        return this.accountAddressRestrictionTransactionBody.getRestrictionDeletions();
    };
    AccountAddressRestrictionTransactionBuilder.prototype.getSize = function () {
        var size = _super.prototype.getSize.call(this);
        size += this.accountAddressRestrictionTransactionBody.getSize();
        return size;
    };
    AccountAddressRestrictionTransactionBuilder.prototype.serialize = function () {
        var newArray = Uint8Array.from([]);
        var superBytes = _super.prototype.serialize.call(this);
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, superBytes);
        var accountAddressRestrictionTransactionBodyBytes = this.accountAddressRestrictionTransactionBody.serialize();
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, accountAddressRestrictionTransactionBodyBytes);
        return newArray;
    };
    return AccountAddressRestrictionTransactionBuilder;
}(TransactionBuilder_1.TransactionBuilder));
exports.AccountAddressRestrictionTransactionBuilder = AccountAddressRestrictionTransactionBuilder;
