// catbuffer-typescript/builders/AddressResolutionEntryBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var AddressDto_1 = require("./AddressDto");
var GeneratorUtils_1 = require("./GeneratorUtils");
var ReceiptSourceBuilder_1 = require("./ReceiptSourceBuilder");
var AddressResolutionEntryBuilder = (function () {
    function AddressResolutionEntryBuilder(source, resolved) {
        this.source = source;
        this.resolved = resolved;
    }
    AddressResolutionEntryBuilder.loadFromBinary = function (payload) {
        var byteArray = Array.from(payload);
        var source = ReceiptSourceBuilder_1.ReceiptSourceBuilder.loadFromBinary(Uint8Array.from(byteArray));
        byteArray.splice(0, source.getSize());
        var resolved = AddressDto_1.AddressDto.loadFromBinary(Uint8Array.from(byteArray));
        byteArray.splice(0, resolved.getSize());
        return new AddressResolutionEntryBuilder(source, resolved);
    };
    AddressResolutionEntryBuilder.prototype.getSource = function () {
        return this.source;
    };
    AddressResolutionEntryBuilder.prototype.getResolved = function () {
        return this.resolved;
    };
    AddressResolutionEntryBuilder.prototype.getSize = function () {
        var size = 0;
        size += this.source.getSize();
        size += this.resolved.getSize();
        return size;
    };
    AddressResolutionEntryBuilder.prototype.serialize = function () {
        var newArray = Uint8Array.from([]);
        var sourceBytes = this.source.serialize();
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, sourceBytes);
        var resolvedBytes = this.resolved.serialize();
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, resolvedBytes);
        return newArray;
    };
    return AddressResolutionEntryBuilder;
}());
exports.AddressResolutionEntryBuilder = AddressResolutionEntryBuilder;
