// catbuffer-typescript/builders/AddressAliasTransactionBodyBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var AddressDto_1 = require("./AddressDto");
var GeneratorUtils_1 = require("./GeneratorUtils");
var NamespaceIdDto_1 = require("./NamespaceIdDto");
var AddressAliasTransactionBodyBuilder = (function () {
    function AddressAliasTransactionBodyBuilder(namespaceId, address, aliasAction) {
        this.namespaceId = namespaceId;
        this.address = address;
        this.aliasAction = aliasAction;
    }
    AddressAliasTransactionBodyBuilder.loadFromBinary = function (payload) {
        var byteArray = Array.from(payload);
        var namespaceId = NamespaceIdDto_1.NamespaceIdDto.loadFromBinary(Uint8Array.from(byteArray));
        byteArray.splice(0, namespaceId.getSize());
        var address = AddressDto_1.AddressDto.loadFromBinary(Uint8Array.from(byteArray));
        byteArray.splice(0, address.getSize());
        var aliasAction = GeneratorUtils_1.GeneratorUtils.bufferToUint(GeneratorUtils_1.GeneratorUtils.getBytes(Uint8Array.from(byteArray), 1));
        byteArray.splice(0, 1);
        return new AddressAliasTransactionBodyBuilder(namespaceId, address, aliasAction);
    };
    AddressAliasTransactionBodyBuilder.prototype.getNamespaceId = function () {
        return this.namespaceId;
    };
    AddressAliasTransactionBodyBuilder.prototype.getAddress = function () {
        return this.address;
    };
    AddressAliasTransactionBodyBuilder.prototype.getAliasAction = function () {
        return this.aliasAction;
    };
    AddressAliasTransactionBodyBuilder.prototype.getSize = function () {
        var size = 0;
        size += this.namespaceId.getSize();
        size += this.address.getSize();
        size += 1;
        return size;
    };
    AddressAliasTransactionBodyBuilder.prototype.serialize = function () {
        var newArray = Uint8Array.from([]);
        var namespaceIdBytes = this.namespaceId.serialize();
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, namespaceIdBytes);
        var addressBytes = this.address.serialize();
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, addressBytes);
        var aliasActionBytes = GeneratorUtils_1.GeneratorUtils.uintToBuffer(this.aliasAction, 1);
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, aliasActionBytes);
        return newArray;
    };
    return AddressAliasTransactionBodyBuilder;
}());
exports.AddressAliasTransactionBodyBuilder = AddressAliasTransactionBodyBuilder;
