// catbuffer-typescript/builders/EmbeddedMosaicGlobalRestrictionTransactionBuilder.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var EmbeddedTransactionBuilder_1 = require("./EmbeddedTransactionBuilder");
var GeneratorUtils_1 = require("./GeneratorUtils");
var MosaicGlobalRestrictionTransactionBodyBuilder_1 = require("./MosaicGlobalRestrictionTransactionBodyBuilder");
var EmbeddedMosaicGlobalRestrictionTransactionBuilder = (function (_super) {
    __extends(EmbeddedMosaicGlobalRestrictionTransactionBuilder, _super);
    function EmbeddedMosaicGlobalRestrictionTransactionBuilder(signerPublicKey, version, network, type, mosaicId, referenceMosaicId, restrictionKey, previousRestrictionValue, newRestrictionValue, previousRestrictionType, newRestrictionType) {
        var _this = _super.call(this, signerPublicKey, version, network, type) || this;
        _this.mosaicGlobalRestrictionTransactionBody = new MosaicGlobalRestrictionTransactionBodyBuilder_1.MosaicGlobalRestrictionTransactionBodyBuilder(mosaicId, referenceMosaicId, restrictionKey, previousRestrictionValue, newRestrictionValue, previousRestrictionType, newRestrictionType);
        return _this;
    }
    EmbeddedMosaicGlobalRestrictionTransactionBuilder.loadFromBinary = function (payload) {
        var byteArray = Array.from(payload);
        var superObject = EmbeddedTransactionBuilder_1.EmbeddedTransactionBuilder.loadFromBinary(Uint8Array.from(byteArray));
        byteArray.splice(0, superObject.getSize());
        var mosaicGlobalRestrictionTransactionBody = MosaicGlobalRestrictionTransactionBodyBuilder_1.MosaicGlobalRestrictionTransactionBodyBuilder.loadFromBinary(Uint8Array.from(byteArray));
        byteArray.splice(0, mosaicGlobalRestrictionTransactionBody.getSize());
        return new EmbeddedMosaicGlobalRestrictionTransactionBuilder(superObject.signerPublicKey, superObject.version, superObject.network, superObject.type, mosaicGlobalRestrictionTransactionBody.mosaicId, mosaicGlobalRestrictionTransactionBody.referenceMosaicId, mosaicGlobalRestrictionTransactionBody.restrictionKey, mosaicGlobalRestrictionTransactionBody.previousRestrictionValue, mosaicGlobalRestrictionTransactionBody.newRestrictionValue, mosaicGlobalRestrictionTransactionBody.previousRestrictionType, mosaicGlobalRestrictionTransactionBody.newRestrictionType);
    };
    EmbeddedMosaicGlobalRestrictionTransactionBuilder.prototype.getMosaicId = function () {
        return this.mosaicGlobalRestrictionTransactionBody.getMosaicId();
    };
    EmbeddedMosaicGlobalRestrictionTransactionBuilder.prototype.getReferenceMosaicId = function () {
        return this.mosaicGlobalRestrictionTransactionBody.getReferenceMosaicId();
    };
    EmbeddedMosaicGlobalRestrictionTransactionBuilder.prototype.getRestrictionKey = function () {
        return this.mosaicGlobalRestrictionTransactionBody.getRestrictionKey();
    };
    EmbeddedMosaicGlobalRestrictionTransactionBuilder.prototype.getPreviousRestrictionValue = function () {
        return this.mosaicGlobalRestrictionTransactionBody.getPreviousRestrictionValue();
    };
    EmbeddedMosaicGlobalRestrictionTransactionBuilder.prototype.getNewRestrictionValue = function () {
        return this.mosaicGlobalRestrictionTransactionBody.getNewRestrictionValue();
    };
    EmbeddedMosaicGlobalRestrictionTransactionBuilder.prototype.getPreviousRestrictionType = function () {
        return this.mosaicGlobalRestrictionTransactionBody.getPreviousRestrictionType();
    };
    EmbeddedMosaicGlobalRestrictionTransactionBuilder.prototype.getNewRestrictionType = function () {
        return this.mosaicGlobalRestrictionTransactionBody.getNewRestrictionType();
    };
    EmbeddedMosaicGlobalRestrictionTransactionBuilder.prototype.getSize = function () {
        var size = _super.prototype.getSize.call(this);
        size += this.mosaicGlobalRestrictionTransactionBody.getSize();
        return size;
    };
    EmbeddedMosaicGlobalRestrictionTransactionBuilder.prototype.serialize = function () {
        var newArray = Uint8Array.from([]);
        var superBytes = _super.prototype.serialize.call(this);
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, superBytes);
        var mosaicGlobalRestrictionTransactionBodyBytes = this.mosaicGlobalRestrictionTransactionBody.serialize();
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, mosaicGlobalRestrictionTransactionBodyBytes);
        return newArray;
    };
    return EmbeddedMosaicGlobalRestrictionTransactionBuilder;
}(EmbeddedTransactionBuilder_1.EmbeddedTransactionBuilder));
exports.EmbeddedMosaicGlobalRestrictionTransactionBuilder = EmbeddedMosaicGlobalRestrictionTransactionBuilder;
