// catbuffer-typescript/builders/EmbeddedMosaicMetadataTransactionBuilder.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var EmbeddedTransactionBuilder_1 = require("./EmbeddedTransactionBuilder");
var GeneratorUtils_1 = require("./GeneratorUtils");
var MosaicMetadataTransactionBodyBuilder_1 = require("./MosaicMetadataTransactionBodyBuilder");
var EmbeddedMosaicMetadataTransactionBuilder = (function (_super) {
    __extends(EmbeddedMosaicMetadataTransactionBuilder, _super);
    function EmbeddedMosaicMetadataTransactionBuilder(signerPublicKey, version, network, type, targetPublicKey, scopedMetadataKey, targetMosaicId, valueSizeDelta, value) {
        var _this = _super.call(this, signerPublicKey, version, network, type) || this;
        _this.mosaicMetadataTransactionBody = new MosaicMetadataTransactionBodyBuilder_1.MosaicMetadataTransactionBodyBuilder(targetPublicKey, scopedMetadataKey, targetMosaicId, valueSizeDelta, value);
        return _this;
    }
    EmbeddedMosaicMetadataTransactionBuilder.loadFromBinary = function (payload) {
        var byteArray = Array.from(payload);
        var superObject = EmbeddedTransactionBuilder_1.EmbeddedTransactionBuilder.loadFromBinary(Uint8Array.from(byteArray));
        byteArray.splice(0, superObject.getSize());
        var mosaicMetadataTransactionBody = MosaicMetadataTransactionBodyBuilder_1.MosaicMetadataTransactionBodyBuilder.loadFromBinary(Uint8Array.from(byteArray));
        byteArray.splice(0, mosaicMetadataTransactionBody.getSize());
        return new EmbeddedMosaicMetadataTransactionBuilder(superObject.signerPublicKey, superObject.version, superObject.network, superObject.type, mosaicMetadataTransactionBody.targetPublicKey, mosaicMetadataTransactionBody.scopedMetadataKey, mosaicMetadataTransactionBody.targetMosaicId, mosaicMetadataTransactionBody.valueSizeDelta, mosaicMetadataTransactionBody.value);
    };
    EmbeddedMosaicMetadataTransactionBuilder.prototype.getTargetPublicKey = function () {
        return this.mosaicMetadataTransactionBody.getTargetPublicKey();
    };
    EmbeddedMosaicMetadataTransactionBuilder.prototype.getScopedMetadataKey = function () {
        return this.mosaicMetadataTransactionBody.getScopedMetadataKey();
    };
    EmbeddedMosaicMetadataTransactionBuilder.prototype.getTargetMosaicId = function () {
        return this.mosaicMetadataTransactionBody.getTargetMosaicId();
    };
    EmbeddedMosaicMetadataTransactionBuilder.prototype.getValueSizeDelta = function () {
        return this.mosaicMetadataTransactionBody.getValueSizeDelta();
    };
    EmbeddedMosaicMetadataTransactionBuilder.prototype.getValue = function () {
        return this.mosaicMetadataTransactionBody.getValue();
    };
    EmbeddedMosaicMetadataTransactionBuilder.prototype.getSize = function () {
        var size = _super.prototype.getSize.call(this);
        size += this.mosaicMetadataTransactionBody.getSize();
        return size;
    };
    EmbeddedMosaicMetadataTransactionBuilder.prototype.serialize = function () {
        var newArray = Uint8Array.from([]);
        var superBytes = _super.prototype.serialize.call(this);
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, superBytes);
        var mosaicMetadataTransactionBodyBytes = this.mosaicMetadataTransactionBody.serialize();
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, mosaicMetadataTransactionBodyBytes);
        return newArray;
    };
    return EmbeddedMosaicMetadataTransactionBuilder;
}(EmbeddedTransactionBuilder_1.EmbeddedTransactionBuilder));
exports.EmbeddedMosaicMetadataTransactionBuilder = EmbeddedMosaicMetadataTransactionBuilder;
