// catbuffer-typescript/builders/EmbeddedTransferTransactionBuilder.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var EmbeddedTransactionBuilder_1 = require("./EmbeddedTransactionBuilder");
var GeneratorUtils_1 = require("./GeneratorUtils");
var TransferTransactionBodyBuilder_1 = require("./TransferTransactionBodyBuilder");
var EmbeddedTransferTransactionBuilder = (function (_super) {
    __extends(EmbeddedTransferTransactionBuilder, _super);
    function EmbeddedTransferTransactionBuilder(signerPublicKey, version, network, type, recipientAddress, mosaics, message) {
        var _this = _super.call(this, signerPublicKey, version, network, type) || this;
        _this.transferTransactionBody = new TransferTransactionBodyBuilder_1.TransferTransactionBodyBuilder(recipientAddress, mosaics, message);
        return _this;
    }
    EmbeddedTransferTransactionBuilder.loadFromBinary = function (payload) {
        var byteArray = Array.from(payload);
        var superObject = EmbeddedTransactionBuilder_1.EmbeddedTransactionBuilder.loadFromBinary(Uint8Array.from(byteArray));
        byteArray.splice(0, superObject.getSize());
        var transferTransactionBody = TransferTransactionBodyBuilder_1.TransferTransactionBodyBuilder.loadFromBinary(Uint8Array.from(byteArray));
        byteArray.splice(0, transferTransactionBody.getSize());
        return new EmbeddedTransferTransactionBuilder(superObject.signerPublicKey, superObject.version, superObject.network, superObject.type, transferTransactionBody.recipientAddress, transferTransactionBody.mosaics, transferTransactionBody.message);
    };
    EmbeddedTransferTransactionBuilder.prototype.getRecipientAddress = function () {
        return this.transferTransactionBody.getRecipientAddress();
    };
    EmbeddedTransferTransactionBuilder.prototype.getTransferTransactionBody_Reserved1 = function () {
        return this.transferTransactionBody.getTransferTransactionBody_Reserved1();
    };
    EmbeddedTransferTransactionBuilder.prototype.getMosaics = function () {
        return this.transferTransactionBody.getMosaics();
    };
    EmbeddedTransferTransactionBuilder.prototype.getMessage = function () {
        return this.transferTransactionBody.getMessage();
    };
    EmbeddedTransferTransactionBuilder.prototype.getSize = function () {
        var size = _super.prototype.getSize.call(this);
        size += this.transferTransactionBody.getSize();
        return size;
    };
    EmbeddedTransferTransactionBuilder.prototype.serialize = function () {
        var newArray = Uint8Array.from([]);
        var superBytes = _super.prototype.serialize.call(this);
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, superBytes);
        var transferTransactionBodyBytes = this.transferTransactionBody.serialize();
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, transferTransactionBodyBytes);
        return newArray;
    };
    return EmbeddedTransferTransactionBuilder;
}(EmbeddedTransactionBuilder_1.EmbeddedTransactionBuilder));
exports.EmbeddedTransferTransactionBuilder = EmbeddedTransferTransactionBuilder;
