// catbuffer-typescript/builders/EmbeddedMosaicDefinitionTransactionBuilder.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var EmbeddedTransactionBuilder_1 = require("./EmbeddedTransactionBuilder");
var GeneratorUtils_1 = require("./GeneratorUtils");
var MosaicDefinitionTransactionBodyBuilder_1 = require("./MosaicDefinitionTransactionBodyBuilder");
var EmbeddedMosaicDefinitionTransactionBuilder = (function (_super) {
    __extends(EmbeddedMosaicDefinitionTransactionBuilder, _super);
    function EmbeddedMosaicDefinitionTransactionBuilder(signerPublicKey, version, network, type, id, duration, nonce, flags, divisibility) {
        var _this = _super.call(this, signerPublicKey, version, network, type) || this;
        _this.mosaicDefinitionTransactionBody = new MosaicDefinitionTransactionBodyBuilder_1.MosaicDefinitionTransactionBodyBuilder(id, duration, nonce, flags, divisibility);
        return _this;
    }
    EmbeddedMosaicDefinitionTransactionBuilder.loadFromBinary = function (payload) {
        var byteArray = Array.from(payload);
        var superObject = EmbeddedTransactionBuilder_1.EmbeddedTransactionBuilder.loadFromBinary(Uint8Array.from(byteArray));
        byteArray.splice(0, superObject.getSize());
        var mosaicDefinitionTransactionBody = MosaicDefinitionTransactionBodyBuilder_1.MosaicDefinitionTransactionBodyBuilder.loadFromBinary(Uint8Array.from(byteArray));
        byteArray.splice(0, mosaicDefinitionTransactionBody.getSize());
        return new EmbeddedMosaicDefinitionTransactionBuilder(superObject.signerPublicKey, superObject.version, superObject.network, superObject.type, mosaicDefinitionTransactionBody.id, mosaicDefinitionTransactionBody.duration, mosaicDefinitionTransactionBody.nonce, mosaicDefinitionTransactionBody.flags, mosaicDefinitionTransactionBody.divisibility);
    };
    EmbeddedMosaicDefinitionTransactionBuilder.prototype.getId = function () {
        return this.mosaicDefinitionTransactionBody.getId();
    };
    EmbeddedMosaicDefinitionTransactionBuilder.prototype.getDuration = function () {
        return this.mosaicDefinitionTransactionBody.getDuration();
    };
    EmbeddedMosaicDefinitionTransactionBuilder.prototype.getNonce = function () {
        return this.mosaicDefinitionTransactionBody.getNonce();
    };
    EmbeddedMosaicDefinitionTransactionBuilder.prototype.getFlags = function () {
        return this.mosaicDefinitionTransactionBody.getFlags();
    };
    EmbeddedMosaicDefinitionTransactionBuilder.prototype.getDivisibility = function () {
        return this.mosaicDefinitionTransactionBody.getDivisibility();
    };
    EmbeddedMosaicDefinitionTransactionBuilder.prototype.getSize = function () {
        var size = _super.prototype.getSize.call(this);
        size += this.mosaicDefinitionTransactionBody.getSize();
        return size;
    };
    EmbeddedMosaicDefinitionTransactionBuilder.prototype.serialize = function () {
        var newArray = Uint8Array.from([]);
        var superBytes = _super.prototype.serialize.call(this);
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, superBytes);
        var mosaicDefinitionTransactionBodyBytes = this.mosaicDefinitionTransactionBody.serialize();
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, mosaicDefinitionTransactionBodyBytes);
        return newArray;
    };
    return EmbeddedMosaicDefinitionTransactionBuilder;
}(EmbeddedTransactionBuilder_1.EmbeddedTransactionBuilder));
exports.EmbeddedMosaicDefinitionTransactionBuilder = EmbeddedMosaicDefinitionTransactionBuilder;
