// catbuffer-typescript/builders/EmbeddedSecretLockTransactionBuilder.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var EmbeddedTransactionBuilder_1 = require("./EmbeddedTransactionBuilder");
var GeneratorUtils_1 = require("./GeneratorUtils");
var SecretLockTransactionBodyBuilder_1 = require("./SecretLockTransactionBodyBuilder");
var EmbeddedSecretLockTransactionBuilder = (function (_super) {
    __extends(EmbeddedSecretLockTransactionBuilder, _super);
    function EmbeddedSecretLockTransactionBuilder(signerPublicKey, version, network, type, secret, mosaic, duration, hashAlgorithm, recipientAddress) {
        var _this = _super.call(this, signerPublicKey, version, network, type) || this;
        _this.secretLockTransactionBody = new SecretLockTransactionBodyBuilder_1.SecretLockTransactionBodyBuilder(secret, mosaic, duration, hashAlgorithm, recipientAddress);
        return _this;
    }
    EmbeddedSecretLockTransactionBuilder.loadFromBinary = function (payload) {
        var byteArray = Array.from(payload);
        var superObject = EmbeddedTransactionBuilder_1.EmbeddedTransactionBuilder.loadFromBinary(Uint8Array.from(byteArray));
        byteArray.splice(0, superObject.getSize());
        var secretLockTransactionBody = SecretLockTransactionBodyBuilder_1.SecretLockTransactionBodyBuilder.loadFromBinary(Uint8Array.from(byteArray));
        byteArray.splice(0, secretLockTransactionBody.getSize());
        return new EmbeddedSecretLockTransactionBuilder(superObject.signerPublicKey, superObject.version, superObject.network, superObject.type, secretLockTransactionBody.secret, secretLockTransactionBody.mosaic, secretLockTransactionBody.duration, secretLockTransactionBody.hashAlgorithm, secretLockTransactionBody.recipientAddress);
    };
    EmbeddedSecretLockTransactionBuilder.prototype.getSecret = function () {
        return this.secretLockTransactionBody.getSecret();
    };
    EmbeddedSecretLockTransactionBuilder.prototype.getMosaic = function () {
        return this.secretLockTransactionBody.getMosaic();
    };
    EmbeddedSecretLockTransactionBuilder.prototype.getDuration = function () {
        return this.secretLockTransactionBody.getDuration();
    };
    EmbeddedSecretLockTransactionBuilder.prototype.getHashAlgorithm = function () {
        return this.secretLockTransactionBody.getHashAlgorithm();
    };
    EmbeddedSecretLockTransactionBuilder.prototype.getRecipientAddress = function () {
        return this.secretLockTransactionBody.getRecipientAddress();
    };
    EmbeddedSecretLockTransactionBuilder.prototype.getSize = function () {
        var size = _super.prototype.getSize.call(this);
        size += this.secretLockTransactionBody.getSize();
        return size;
    };
    EmbeddedSecretLockTransactionBuilder.prototype.serialize = function () {
        var newArray = Uint8Array.from([]);
        var superBytes = _super.prototype.serialize.call(this);
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, superBytes);
        var secretLockTransactionBodyBytes = this.secretLockTransactionBody.serialize();
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, secretLockTransactionBodyBytes);
        return newArray;
    };
    return EmbeddedSecretLockTransactionBuilder;
}(EmbeddedTransactionBuilder_1.EmbeddedTransactionBuilder));
exports.EmbeddedSecretLockTransactionBuilder = EmbeddedSecretLockTransactionBuilder;
