// catbuffer-typescript/builders/EmbeddedMosaicAddressRestrictionTransactionBuilder.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var EmbeddedTransactionBuilder_1 = require("./EmbeddedTransactionBuilder");
var GeneratorUtils_1 = require("./GeneratorUtils");
var MosaicAddressRestrictionTransactionBodyBuilder_1 = require("./MosaicAddressRestrictionTransactionBodyBuilder");
var EmbeddedMosaicAddressRestrictionTransactionBuilder = (function (_super) {
    __extends(EmbeddedMosaicAddressRestrictionTransactionBuilder, _super);
    function EmbeddedMosaicAddressRestrictionTransactionBuilder(signerPublicKey, version, network, type, mosaicId, restrictionKey, previousRestrictionValue, newRestrictionValue, targetAddress) {
        var _this = _super.call(this, signerPublicKey, version, network, type) || this;
        _this.mosaicAddressRestrictionTransactionBody = new MosaicAddressRestrictionTransactionBodyBuilder_1.MosaicAddressRestrictionTransactionBodyBuilder(mosaicId, restrictionKey, previousRestrictionValue, newRestrictionValue, targetAddress);
        return _this;
    }
    EmbeddedMosaicAddressRestrictionTransactionBuilder.loadFromBinary = function (payload) {
        var byteArray = Array.from(payload);
        var superObject = EmbeddedTransactionBuilder_1.EmbeddedTransactionBuilder.loadFromBinary(Uint8Array.from(byteArray));
        byteArray.splice(0, superObject.getSize());
        var mosaicAddressRestrictionTransactionBody = MosaicAddressRestrictionTransactionBodyBuilder_1.MosaicAddressRestrictionTransactionBodyBuilder.loadFromBinary(Uint8Array.from(byteArray));
        byteArray.splice(0, mosaicAddressRestrictionTransactionBody.getSize());
        return new EmbeddedMosaicAddressRestrictionTransactionBuilder(superObject.signerPublicKey, superObject.version, superObject.network, superObject.type, mosaicAddressRestrictionTransactionBody.mosaicId, mosaicAddressRestrictionTransactionBody.restrictionKey, mosaicAddressRestrictionTransactionBody.previousRestrictionValue, mosaicAddressRestrictionTransactionBody.newRestrictionValue, mosaicAddressRestrictionTransactionBody.targetAddress);
    };
    EmbeddedMosaicAddressRestrictionTransactionBuilder.prototype.getMosaicId = function () {
        return this.mosaicAddressRestrictionTransactionBody.getMosaicId();
    };
    EmbeddedMosaicAddressRestrictionTransactionBuilder.prototype.getRestrictionKey = function () {
        return this.mosaicAddressRestrictionTransactionBody.getRestrictionKey();
    };
    EmbeddedMosaicAddressRestrictionTransactionBuilder.prototype.getPreviousRestrictionValue = function () {
        return this.mosaicAddressRestrictionTransactionBody.getPreviousRestrictionValue();
    };
    EmbeddedMosaicAddressRestrictionTransactionBuilder.prototype.getNewRestrictionValue = function () {
        return this.mosaicAddressRestrictionTransactionBody.getNewRestrictionValue();
    };
    EmbeddedMosaicAddressRestrictionTransactionBuilder.prototype.getTargetAddress = function () {
        return this.mosaicAddressRestrictionTransactionBody.getTargetAddress();
    };
    EmbeddedMosaicAddressRestrictionTransactionBuilder.prototype.getSize = function () {
        var size = _super.prototype.getSize.call(this);
        size += this.mosaicAddressRestrictionTransactionBody.getSize();
        return size;
    };
    EmbeddedMosaicAddressRestrictionTransactionBuilder.prototype.serialize = function () {
        var newArray = Uint8Array.from([]);
        var superBytes = _super.prototype.serialize.call(this);
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, superBytes);
        var mosaicAddressRestrictionTransactionBodyBytes = this.mosaicAddressRestrictionTransactionBody.serialize();
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, mosaicAddressRestrictionTransactionBodyBytes);
        return newArray;
    };
    return EmbeddedMosaicAddressRestrictionTransactionBuilder;
}(EmbeddedTransactionBuilder_1.EmbeddedTransactionBuilder));
exports.EmbeddedMosaicAddressRestrictionTransactionBuilder = EmbeddedMosaicAddressRestrictionTransactionBuilder;
