// catbuffer-typescript/builders/EmbeddedHashLockTransactionBuilder.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var EmbeddedTransactionBuilder_1 = require("./EmbeddedTransactionBuilder");
var GeneratorUtils_1 = require("./GeneratorUtils");
var HashLockTransactionBodyBuilder_1 = require("./HashLockTransactionBodyBuilder");
var EmbeddedHashLockTransactionBuilder = (function (_super) {
    __extends(EmbeddedHashLockTransactionBuilder, _super);
    function EmbeddedHashLockTransactionBuilder(signerPublicKey, version, network, type, mosaic, duration, hash) {
        var _this = _super.call(this, signerPublicKey, version, network, type) || this;
        _this.hashLockTransactionBody = new HashLockTransactionBodyBuilder_1.HashLockTransactionBodyBuilder(mosaic, duration, hash);
        return _this;
    }
    EmbeddedHashLockTransactionBuilder.loadFromBinary = function (payload) {
        var byteArray = Array.from(payload);
        var superObject = EmbeddedTransactionBuilder_1.EmbeddedTransactionBuilder.loadFromBinary(Uint8Array.from(byteArray));
        byteArray.splice(0, superObject.getSize());
        var hashLockTransactionBody = HashLockTransactionBodyBuilder_1.HashLockTransactionBodyBuilder.loadFromBinary(Uint8Array.from(byteArray));
        byteArray.splice(0, hashLockTransactionBody.getSize());
        return new EmbeddedHashLockTransactionBuilder(superObject.signerPublicKey, superObject.version, superObject.network, superObject.type, hashLockTransactionBody.mosaic, hashLockTransactionBody.duration, hashLockTransactionBody.hash);
    };
    EmbeddedHashLockTransactionBuilder.prototype.getMosaic = function () {
        return this.hashLockTransactionBody.getMosaic();
    };
    EmbeddedHashLockTransactionBuilder.prototype.getDuration = function () {
        return this.hashLockTransactionBody.getDuration();
    };
    EmbeddedHashLockTransactionBuilder.prototype.getHash = function () {
        return this.hashLockTransactionBody.getHash();
    };
    EmbeddedHashLockTransactionBuilder.prototype.getSize = function () {
        var size = _super.prototype.getSize.call(this);
        size += this.hashLockTransactionBody.getSize();
        return size;
    };
    EmbeddedHashLockTransactionBuilder.prototype.serialize = function () {
        var newArray = Uint8Array.from([]);
        var superBytes = _super.prototype.serialize.call(this);
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, superBytes);
        var hashLockTransactionBodyBytes = this.hashLockTransactionBody.serialize();
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, hashLockTransactionBodyBytes);
        return newArray;
    };
    return EmbeddedHashLockTransactionBuilder;
}(EmbeddedTransactionBuilder_1.EmbeddedTransactionBuilder));
exports.EmbeddedHashLockTransactionBuilder = EmbeddedHashLockTransactionBuilder;
