// catbuffer-typescript/builders/EmbeddedSecretProofTransactionBuilder.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var EmbeddedTransactionBuilder_1 = require("./EmbeddedTransactionBuilder");
var GeneratorUtils_1 = require("./GeneratorUtils");
var SecretProofTransactionBodyBuilder_1 = require("./SecretProofTransactionBodyBuilder");
var EmbeddedSecretProofTransactionBuilder = (function (_super) {
    __extends(EmbeddedSecretProofTransactionBuilder, _super);
    function EmbeddedSecretProofTransactionBuilder(signerPublicKey, version, network, type, secret, hashAlgorithm, recipientAddress, proof) {
        var _this = _super.call(this, signerPublicKey, version, network, type) || this;
        _this.secretProofTransactionBody = new SecretProofTransactionBodyBuilder_1.SecretProofTransactionBodyBuilder(secret, hashAlgorithm, recipientAddress, proof);
        return _this;
    }
    EmbeddedSecretProofTransactionBuilder.loadFromBinary = function (payload) {
        var byteArray = Array.from(payload);
        var superObject = EmbeddedTransactionBuilder_1.EmbeddedTransactionBuilder.loadFromBinary(Uint8Array.from(byteArray));
        byteArray.splice(0, superObject.getSize());
        var secretProofTransactionBody = SecretProofTransactionBodyBuilder_1.SecretProofTransactionBodyBuilder.loadFromBinary(Uint8Array.from(byteArray));
        byteArray.splice(0, secretProofTransactionBody.getSize());
        return new EmbeddedSecretProofTransactionBuilder(superObject.signerPublicKey, superObject.version, superObject.network, superObject.type, secretProofTransactionBody.secret, secretProofTransactionBody.hashAlgorithm, secretProofTransactionBody.recipientAddress, secretProofTransactionBody.proof);
    };
    EmbeddedSecretProofTransactionBuilder.prototype.getSecret = function () {
        return this.secretProofTransactionBody.getSecret();
    };
    EmbeddedSecretProofTransactionBuilder.prototype.getHashAlgorithm = function () {
        return this.secretProofTransactionBody.getHashAlgorithm();
    };
    EmbeddedSecretProofTransactionBuilder.prototype.getRecipientAddress = function () {
        return this.secretProofTransactionBody.getRecipientAddress();
    };
    EmbeddedSecretProofTransactionBuilder.prototype.getProof = function () {
        return this.secretProofTransactionBody.getProof();
    };
    EmbeddedSecretProofTransactionBuilder.prototype.getSize = function () {
        var size = _super.prototype.getSize.call(this);
        size += this.secretProofTransactionBody.getSize();
        return size;
    };
    EmbeddedSecretProofTransactionBuilder.prototype.serialize = function () {
        var newArray = Uint8Array.from([]);
        var superBytes = _super.prototype.serialize.call(this);
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, superBytes);
        var secretProofTransactionBodyBytes = this.secretProofTransactionBody.serialize();
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, secretProofTransactionBodyBytes);
        return newArray;
    };
    return EmbeddedSecretProofTransactionBuilder;
}(EmbeddedTransactionBuilder_1.EmbeddedTransactionBuilder));
exports.EmbeddedSecretProofTransactionBuilder = EmbeddedSecretProofTransactionBuilder;
