// catbuffer-typescript/builders/HeightActivityBucketBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var AmountDto_1 = require("./AmountDto");
var GeneratorUtils_1 = require("./GeneratorUtils");
var ImportanceHeightDto_1 = require("./ImportanceHeightDto");
var HeightActivityBucketBuilder = (function () {
    function HeightActivityBucketBuilder(startHeight, totalFeesPaid, beneficiaryCount, rawScore) {
        this.startHeight = startHeight;
        this.totalFeesPaid = totalFeesPaid;
        this.beneficiaryCount = beneficiaryCount;
        this.rawScore = rawScore;
    }
    HeightActivityBucketBuilder.loadFromBinary = function (payload) {
        var byteArray = Array.from(payload);
        var startHeight = ImportanceHeightDto_1.ImportanceHeightDto.loadFromBinary(Uint8Array.from(byteArray));
        byteArray.splice(0, startHeight.getSize());
        var totalFeesPaid = AmountDto_1.AmountDto.loadFromBinary(Uint8Array.from(byteArray));
        byteArray.splice(0, totalFeesPaid.getSize());
        var beneficiaryCount = GeneratorUtils_1.GeneratorUtils.bufferToUint(GeneratorUtils_1.GeneratorUtils.getBytes(Uint8Array.from(byteArray), 4));
        byteArray.splice(0, 4);
        var rawScore = GeneratorUtils_1.GeneratorUtils.bufferToUint64(GeneratorUtils_1.GeneratorUtils.getBytes(Uint8Array.from(byteArray), 8));
        byteArray.splice(0, 8);
        return new HeightActivityBucketBuilder(startHeight, totalFeesPaid, beneficiaryCount, rawScore);
    };
    HeightActivityBucketBuilder.prototype.getStartHeight = function () {
        return this.startHeight;
    };
    HeightActivityBucketBuilder.prototype.getTotalFeesPaid = function () {
        return this.totalFeesPaid;
    };
    HeightActivityBucketBuilder.prototype.getBeneficiaryCount = function () {
        return this.beneficiaryCount;
    };
    HeightActivityBucketBuilder.prototype.getRawScore = function () {
        return this.rawScore;
    };
    HeightActivityBucketBuilder.prototype.getSize = function () {
        var size = 0;
        size += this.startHeight.getSize();
        size += this.totalFeesPaid.getSize();
        size += 4;
        size += 8;
        return size;
    };
    HeightActivityBucketBuilder.prototype.serialize = function () {
        var newArray = Uint8Array.from([]);
        var startHeightBytes = this.startHeight.serialize();
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, startHeightBytes);
        var totalFeesPaidBytes = this.totalFeesPaid.serialize();
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, totalFeesPaidBytes);
        var beneficiaryCountBytes = GeneratorUtils_1.GeneratorUtils.uintToBuffer(this.getBeneficiaryCount(), 4);
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, beneficiaryCountBytes);
        var rawScoreBytes = GeneratorUtils_1.GeneratorUtils.uint64ToBuffer(this.getRawScore());
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, rawScoreBytes);
        return newArray;
    };
    return HeightActivityBucketBuilder;
}());
exports.HeightActivityBucketBuilder = HeightActivityBucketBuilder;
