// catbuffer-typescript/builders/ImportanceSnapshotBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var GeneratorUtils_1 = require("./GeneratorUtils");
var ImportanceDto_1 = require("./ImportanceDto");
var ImportanceHeightDto_1 = require("./ImportanceHeightDto");
var ImportanceSnapshotBuilder = (function () {
    function ImportanceSnapshotBuilder(importance, height) {
        this.importance = importance;
        this.height = height;
    }
    ImportanceSnapshotBuilder.loadFromBinary = function (payload) {
        var byteArray = Array.from(payload);
        var importance = ImportanceDto_1.ImportanceDto.loadFromBinary(Uint8Array.from(byteArray));
        byteArray.splice(0, importance.getSize());
        var height = ImportanceHeightDto_1.ImportanceHeightDto.loadFromBinary(Uint8Array.from(byteArray));
        byteArray.splice(0, height.getSize());
        return new ImportanceSnapshotBuilder(importance, height);
    };
    ImportanceSnapshotBuilder.prototype.getImportance = function () {
        return this.importance;
    };
    ImportanceSnapshotBuilder.prototype.getHeight = function () {
        return this.height;
    };
    ImportanceSnapshotBuilder.prototype.getSize = function () {
        var size = 0;
        size += this.importance.getSize();
        size += this.height.getSize();
        return size;
    };
    ImportanceSnapshotBuilder.prototype.serialize = function () {
        var newArray = Uint8Array.from([]);
        var importanceBytes = this.importance.serialize();
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, importanceBytes);
        var heightBytes = this.height.serialize();
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, heightBytes);
        return newArray;
    };
    return ImportanceSnapshotBuilder;
}());
exports.ImportanceSnapshotBuilder = ImportanceSnapshotBuilder;
