// catbuffer-typescript/builders/MetadataValueBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var GeneratorUtils_1 = require("./GeneratorUtils");
var MetadataValueBuilder = (function () {
    function MetadataValueBuilder(data) {
        this.size = 0;
        this.data = data;
    }
    MetadataValueBuilder.loadFromBinary = function (payload) {
        var byteArray = Array.from(payload);
        var size = GeneratorUtils_1.GeneratorUtils.bufferToUint(GeneratorUtils_1.GeneratorUtils.getBytes(Uint8Array.from(byteArray), 2));
        byteArray.splice(0, 2);
        var data = GeneratorUtils_1.GeneratorUtils.getBytes(Uint8Array.from(byteArray), size);
        byteArray.splice(0, size);
        return new MetadataValueBuilder(data);
    };
    MetadataValueBuilder.prototype.getData = function () {
        return this.data;
    };
    MetadataValueBuilder.prototype.getSize = function () {
        var size = 0;
        size += 2;
        size += this.data.length;
        return size;
    };
    MetadataValueBuilder.prototype.serialize = function () {
        var newArray = Uint8Array.from([]);
        var sizeBytes = GeneratorUtils_1.GeneratorUtils.uintToBuffer(this.getSize(), 2);
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, sizeBytes);
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, this.data);
        return newArray;
    };
    return MetadataValueBuilder;
}());
exports.MetadataValueBuilder = MetadataValueBuilder;
