// catbuffer-typescript/builders/MosaicDefinitionBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var GeneratorUtils_1 = require("./GeneratorUtils");
var HeightDto_1 = require("./HeightDto");
var KeyDto_1 = require("./KeyDto");
var MosaicPropertiesBuilder_1 = require("./MosaicPropertiesBuilder");
var MosaicDefinitionBuilder = (function () {
    function MosaicDefinitionBuilder(startHeight, ownerPublicKey, revision, properties) {
        this.startHeight = startHeight;
        this.ownerPublicKey = ownerPublicKey;
        this.revision = revision;
        this.properties = properties;
    }
    MosaicDefinitionBuilder.loadFromBinary = function (payload) {
        var byteArray = Array.from(payload);
        var startHeight = HeightDto_1.HeightDto.loadFromBinary(Uint8Array.from(byteArray));
        byteArray.splice(0, startHeight.getSize());
        var ownerPublicKey = KeyDto_1.KeyDto.loadFromBinary(Uint8Array.from(byteArray));
        byteArray.splice(0, ownerPublicKey.getSize());
        var revision = GeneratorUtils_1.GeneratorUtils.bufferToUint(GeneratorUtils_1.GeneratorUtils.getBytes(Uint8Array.from(byteArray), 4));
        byteArray.splice(0, 4);
        var properties = MosaicPropertiesBuilder_1.MosaicPropertiesBuilder.loadFromBinary(Uint8Array.from(byteArray));
        byteArray.splice(0, properties.getSize());
        return new MosaicDefinitionBuilder(startHeight, ownerPublicKey, revision, properties);
    };
    MosaicDefinitionBuilder.prototype.getStartHeight = function () {
        return this.startHeight;
    };
    MosaicDefinitionBuilder.prototype.getOwnerPublicKey = function () {
        return this.ownerPublicKey;
    };
    MosaicDefinitionBuilder.prototype.getRevision = function () {
        return this.revision;
    };
    MosaicDefinitionBuilder.prototype.getProperties = function () {
        return this.properties;
    };
    MosaicDefinitionBuilder.prototype.getSize = function () {
        var size = 0;
        size += this.startHeight.getSize();
        size += this.ownerPublicKey.getSize();
        size += 4;
        size += this.properties.getSize();
        return size;
    };
    MosaicDefinitionBuilder.prototype.serialize = function () {
        var newArray = Uint8Array.from([]);
        var startHeightBytes = this.startHeight.serialize();
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, startHeightBytes);
        var ownerPublicKeyBytes = this.ownerPublicKey.serialize();
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, ownerPublicKeyBytes);
        var revisionBytes = GeneratorUtils_1.GeneratorUtils.uintToBuffer(this.getRevision(), 4);
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, revisionBytes);
        var propertiesBytes = this.properties.serialize();
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, propertiesBytes);
        return newArray;
    };
    return MosaicDefinitionBuilder;
}());
exports.MosaicDefinitionBuilder = MosaicDefinitionBuilder;
