// catbuffer-typescript/builders/MosaicBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var AmountDto_1 = require("./AmountDto");
var GeneratorUtils_1 = require("./GeneratorUtils");
var MosaicIdDto_1 = require("./MosaicIdDto");
var MosaicBuilder = (function () {
    function MosaicBuilder(mosaicId, amount) {
        this.mosaicId = mosaicId;
        this.amount = amount;
    }
    MosaicBuilder.loadFromBinary = function (payload) {
        var byteArray = Array.from(payload);
        var mosaicId = MosaicIdDto_1.MosaicIdDto.loadFromBinary(Uint8Array.from(byteArray));
        byteArray.splice(0, mosaicId.getSize());
        var amount = AmountDto_1.AmountDto.loadFromBinary(Uint8Array.from(byteArray));
        byteArray.splice(0, amount.getSize());
        return new MosaicBuilder(mosaicId, amount);
    };
    MosaicBuilder.prototype.getMosaicId = function () {
        return this.mosaicId;
    };
    MosaicBuilder.prototype.getAmount = function () {
        return this.amount;
    };
    MosaicBuilder.prototype.getSize = function () {
        var size = 0;
        size += this.mosaicId.getSize();
        size += this.amount.getSize();
        return size;
    };
    MosaicBuilder.prototype.serialize = function () {
        var newArray = Uint8Array.from([]);
        var mosaicIdBytes = this.mosaicId.serialize();
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, mosaicIdBytes);
        var amountBytes = this.amount.serialize();
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, amountBytes);
        return newArray;
    };
    return MosaicBuilder;
}());
exports.MosaicBuilder = MosaicBuilder;
