// catbuffer-typescript/builders/MosaicGlobalRestrictionEntryBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var GeneratorUtils_1 = require("./GeneratorUtils");
var GlobalKeyValueSetBuilder_1 = require("./GlobalKeyValueSetBuilder");
var MosaicIdDto_1 = require("./MosaicIdDto");
var MosaicGlobalRestrictionEntryBuilder = (function () {
    function MosaicGlobalRestrictionEntryBuilder(mosaicId, keyPairs) {
        this.mosaicId = mosaicId;
        this.keyPairs = keyPairs;
    }
    MosaicGlobalRestrictionEntryBuilder.loadFromBinary = function (payload) {
        var byteArray = Array.from(payload);
        var mosaicId = MosaicIdDto_1.MosaicIdDto.loadFromBinary(Uint8Array.from(byteArray));
        byteArray.splice(0, mosaicId.getSize());
        var keyPairs = GlobalKeyValueSetBuilder_1.GlobalKeyValueSetBuilder.loadFromBinary(Uint8Array.from(byteArray));
        byteArray.splice(0, keyPairs.getSize());
        return new MosaicGlobalRestrictionEntryBuilder(mosaicId, keyPairs);
    };
    MosaicGlobalRestrictionEntryBuilder.prototype.getMosaicId = function () {
        return this.mosaicId;
    };
    MosaicGlobalRestrictionEntryBuilder.prototype.getKeyPairs = function () {
        return this.keyPairs;
    };
    MosaicGlobalRestrictionEntryBuilder.prototype.getSize = function () {
        var size = 0;
        size += this.mosaicId.getSize();
        size += this.keyPairs.getSize();
        return size;
    };
    MosaicGlobalRestrictionEntryBuilder.prototype.serialize = function () {
        var newArray = Uint8Array.from([]);
        var mosaicIdBytes = this.mosaicId.serialize();
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, mosaicIdBytes);
        var keyPairsBytes = this.keyPairs.serialize();
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, keyPairsBytes);
        return newArray;
    };
    return MosaicGlobalRestrictionEntryBuilder;
}());
exports.MosaicGlobalRestrictionEntryBuilder = MosaicGlobalRestrictionEntryBuilder;
