// catbuffer-typescript/builders/MultisigAccountModificationTransactionBuilder.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var GeneratorUtils_1 = require("./GeneratorUtils");
var MultisigAccountModificationTransactionBodyBuilder_1 = require("./MultisigAccountModificationTransactionBodyBuilder");
var TransactionBuilder_1 = require("./TransactionBuilder");
var MultisigAccountModificationTransactionBuilder = (function (_super) {
    __extends(MultisigAccountModificationTransactionBuilder, _super);
    function MultisigAccountModificationTransactionBuilder(signature, signerPublicKey, version, network, type, fee, deadline, minRemovalDelta, minApprovalDelta, publicKeyAdditions, publicKeyDeletions) {
        var _this = _super.call(this, signature, signerPublicKey, version, network, type, fee, deadline) || this;
        _this.multisigAccountModificationTransactionBody = new MultisigAccountModificationTransactionBodyBuilder_1.MultisigAccountModificationTransactionBodyBuilder(minRemovalDelta, minApprovalDelta, publicKeyAdditions, publicKeyDeletions);
        return _this;
    }
    MultisigAccountModificationTransactionBuilder.loadFromBinary = function (payload) {
        var byteArray = Array.from(payload);
        var superObject = TransactionBuilder_1.TransactionBuilder.loadFromBinary(Uint8Array.from(byteArray));
        byteArray.splice(0, superObject.getSize());
        var multisigAccountModificationTransactionBody = MultisigAccountModificationTransactionBodyBuilder_1.MultisigAccountModificationTransactionBodyBuilder.loadFromBinary(Uint8Array.from(byteArray));
        byteArray.splice(0, multisigAccountModificationTransactionBody.getSize());
        return new MultisigAccountModificationTransactionBuilder(superObject.signature, superObject.signerPublicKey, superObject.version, superObject.network, superObject.type, superObject.fee, superObject.deadline, multisigAccountModificationTransactionBody.minRemovalDelta, multisigAccountModificationTransactionBody.minApprovalDelta, multisigAccountModificationTransactionBody.publicKeyAdditions, multisigAccountModificationTransactionBody.publicKeyDeletions);
    };
    MultisigAccountModificationTransactionBuilder.prototype.getMinRemovalDelta = function () {
        return this.multisigAccountModificationTransactionBody.getMinRemovalDelta();
    };
    MultisigAccountModificationTransactionBuilder.prototype.getMinApprovalDelta = function () {
        return this.multisigAccountModificationTransactionBody.getMinApprovalDelta();
    };
    MultisigAccountModificationTransactionBuilder.prototype.getMultisigAccountModificationTransactionBody_Reserved1 = function () {
        return this.multisigAccountModificationTransactionBody.getMultisigAccountModificationTransactionBody_Reserved1();
    };
    MultisigAccountModificationTransactionBuilder.prototype.getPublicKeyAdditions = function () {
        return this.multisigAccountModificationTransactionBody.getPublicKeyAdditions();
    };
    MultisigAccountModificationTransactionBuilder.prototype.getPublicKeyDeletions = function () {
        return this.multisigAccountModificationTransactionBody.getPublicKeyDeletions();
    };
    MultisigAccountModificationTransactionBuilder.prototype.getSize = function () {
        var size = _super.prototype.getSize.call(this);
        size += this.multisigAccountModificationTransactionBody.getSize();
        return size;
    };
    MultisigAccountModificationTransactionBuilder.prototype.serialize = function () {
        var newArray = Uint8Array.from([]);
        var superBytes = _super.prototype.serialize.call(this);
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, superBytes);
        var multisigAccountModificationTransactionBodyBytes = this.multisigAccountModificationTransactionBody.serialize();
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, multisigAccountModificationTransactionBodyBytes);
        return newArray;
    };
    return MultisigAccountModificationTransactionBuilder;
}(TransactionBuilder_1.TransactionBuilder));
exports.MultisigAccountModificationTransactionBuilder = MultisigAccountModificationTransactionBuilder;
