// catbuffer-typescript/builders/NamespaceMetadataTransactionBuilder.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var GeneratorUtils_1 = require("./GeneratorUtils");
var NamespaceMetadataTransactionBodyBuilder_1 = require("./NamespaceMetadataTransactionBodyBuilder");
var TransactionBuilder_1 = require("./TransactionBuilder");
var NamespaceMetadataTransactionBuilder = (function (_super) {
    __extends(NamespaceMetadataTransactionBuilder, _super);
    function NamespaceMetadataTransactionBuilder(signature, signerPublicKey, version, network, type, fee, deadline, targetPublicKey, scopedMetadataKey, targetNamespaceId, valueSizeDelta, value) {
        var _this = _super.call(this, signature, signerPublicKey, version, network, type, fee, deadline) || this;
        _this.namespaceMetadataTransactionBody = new NamespaceMetadataTransactionBodyBuilder_1.NamespaceMetadataTransactionBodyBuilder(targetPublicKey, scopedMetadataKey, targetNamespaceId, valueSizeDelta, value);
        return _this;
    }
    NamespaceMetadataTransactionBuilder.loadFromBinary = function (payload) {
        var byteArray = Array.from(payload);
        var superObject = TransactionBuilder_1.TransactionBuilder.loadFromBinary(Uint8Array.from(byteArray));
        byteArray.splice(0, superObject.getSize());
        var namespaceMetadataTransactionBody = NamespaceMetadataTransactionBodyBuilder_1.NamespaceMetadataTransactionBodyBuilder.loadFromBinary(Uint8Array.from(byteArray));
        byteArray.splice(0, namespaceMetadataTransactionBody.getSize());
        return new NamespaceMetadataTransactionBuilder(superObject.signature, superObject.signerPublicKey, superObject.version, superObject.network, superObject.type, superObject.fee, superObject.deadline, namespaceMetadataTransactionBody.targetPublicKey, namespaceMetadataTransactionBody.scopedMetadataKey, namespaceMetadataTransactionBody.targetNamespaceId, namespaceMetadataTransactionBody.valueSizeDelta, namespaceMetadataTransactionBody.value);
    };
    NamespaceMetadataTransactionBuilder.prototype.getTargetPublicKey = function () {
        return this.namespaceMetadataTransactionBody.getTargetPublicKey();
    };
    NamespaceMetadataTransactionBuilder.prototype.getScopedMetadataKey = function () {
        return this.namespaceMetadataTransactionBody.getScopedMetadataKey();
    };
    NamespaceMetadataTransactionBuilder.prototype.getTargetNamespaceId = function () {
        return this.namespaceMetadataTransactionBody.getTargetNamespaceId();
    };
    NamespaceMetadataTransactionBuilder.prototype.getValueSizeDelta = function () {
        return this.namespaceMetadataTransactionBody.getValueSizeDelta();
    };
    NamespaceMetadataTransactionBuilder.prototype.getValue = function () {
        return this.namespaceMetadataTransactionBody.getValue();
    };
    NamespaceMetadataTransactionBuilder.prototype.getSize = function () {
        var size = _super.prototype.getSize.call(this);
        size += this.namespaceMetadataTransactionBody.getSize();
        return size;
    };
    NamespaceMetadataTransactionBuilder.prototype.serialize = function () {
        var newArray = Uint8Array.from([]);
        var superBytes = _super.prototype.serialize.call(this);
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, superBytes);
        var namespaceMetadataTransactionBodyBytes = this.namespaceMetadataTransactionBody.serialize();
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, namespaceMetadataTransactionBodyBytes);
        return newArray;
    };
    return NamespaceMetadataTransactionBuilder;
}(TransactionBuilder_1.TransactionBuilder));
exports.NamespaceMetadataTransactionBuilder = NamespaceMetadataTransactionBuilder;
