// catbuffer-typescript/builders/NamespaceLifetimeBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var GeneratorUtils_1 = require("./GeneratorUtils");
var HeightDto_1 = require("./HeightDto");
var NamespaceLifetimeBuilder = (function () {
    function NamespaceLifetimeBuilder(lifetimeStart, lifetimeEnd) {
        this.lifetimeStart = lifetimeStart;
        this.lifetimeEnd = lifetimeEnd;
    }
    NamespaceLifetimeBuilder.loadFromBinary = function (payload) {
        var byteArray = Array.from(payload);
        var lifetimeStart = HeightDto_1.HeightDto.loadFromBinary(Uint8Array.from(byteArray));
        byteArray.splice(0, lifetimeStart.getSize());
        var lifetimeEnd = HeightDto_1.HeightDto.loadFromBinary(Uint8Array.from(byteArray));
        byteArray.splice(0, lifetimeEnd.getSize());
        return new NamespaceLifetimeBuilder(lifetimeStart, lifetimeEnd);
    };
    NamespaceLifetimeBuilder.prototype.getLifetimeStart = function () {
        return this.lifetimeStart;
    };
    NamespaceLifetimeBuilder.prototype.getLifetimeEnd = function () {
        return this.lifetimeEnd;
    };
    NamespaceLifetimeBuilder.prototype.getSize = function () {
        var size = 0;
        size += this.lifetimeStart.getSize();
        size += this.lifetimeEnd.getSize();
        return size;
    };
    NamespaceLifetimeBuilder.prototype.serialize = function () {
        var newArray = Uint8Array.from([]);
        var lifetimeStartBytes = this.lifetimeStart.serialize();
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, lifetimeStartBytes);
        var lifetimeEndBytes = this.lifetimeEnd.serialize();
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, lifetimeEndBytes);
        return newArray;
    };
    return NamespaceLifetimeBuilder;
}());
exports.NamespaceLifetimeBuilder = NamespaceLifetimeBuilder;
