// catbuffer-typescript/builders/SecretProofTransactionBuilder.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var GeneratorUtils_1 = require("./GeneratorUtils");
var SecretProofTransactionBodyBuilder_1 = require("./SecretProofTransactionBodyBuilder");
var TransactionBuilder_1 = require("./TransactionBuilder");
var SecretProofTransactionBuilder = (function (_super) {
    __extends(SecretProofTransactionBuilder, _super);
    function SecretProofTransactionBuilder(signature, signerPublicKey, version, network, type, fee, deadline, secret, hashAlgorithm, recipientAddress, proof) {
        var _this = _super.call(this, signature, signerPublicKey, version, network, type, fee, deadline) || this;
        _this.secretProofTransactionBody = new SecretProofTransactionBodyBuilder_1.SecretProofTransactionBodyBuilder(secret, hashAlgorithm, recipientAddress, proof);
        return _this;
    }
    SecretProofTransactionBuilder.loadFromBinary = function (payload) {
        var byteArray = Array.from(payload);
        var superObject = TransactionBuilder_1.TransactionBuilder.loadFromBinary(Uint8Array.from(byteArray));
        byteArray.splice(0, superObject.getSize());
        var secretProofTransactionBody = SecretProofTransactionBodyBuilder_1.SecretProofTransactionBodyBuilder.loadFromBinary(Uint8Array.from(byteArray));
        byteArray.splice(0, secretProofTransactionBody.getSize());
        return new SecretProofTransactionBuilder(superObject.signature, superObject.signerPublicKey, superObject.version, superObject.network, superObject.type, superObject.fee, superObject.deadline, secretProofTransactionBody.secret, secretProofTransactionBody.hashAlgorithm, secretProofTransactionBody.recipientAddress, secretProofTransactionBody.proof);
    };
    SecretProofTransactionBuilder.prototype.getSecret = function () {
        return this.secretProofTransactionBody.getSecret();
    };
    SecretProofTransactionBuilder.prototype.getHashAlgorithm = function () {
        return this.secretProofTransactionBody.getHashAlgorithm();
    };
    SecretProofTransactionBuilder.prototype.getRecipientAddress = function () {
        return this.secretProofTransactionBody.getRecipientAddress();
    };
    SecretProofTransactionBuilder.prototype.getProof = function () {
        return this.secretProofTransactionBody.getProof();
    };
    SecretProofTransactionBuilder.prototype.getSize = function () {
        var size = _super.prototype.getSize.call(this);
        size += this.secretProofTransactionBody.getSize();
        return size;
    };
    SecretProofTransactionBuilder.prototype.serialize = function () {
        var newArray = Uint8Array.from([]);
        var superBytes = _super.prototype.serialize.call(this);
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, superBytes);
        var secretProofTransactionBodyBytes = this.secretProofTransactionBody.serialize();
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, secretProofTransactionBodyBytes);
        return newArray;
    };
    return SecretProofTransactionBuilder;
}(TransactionBuilder_1.TransactionBuilder));
exports.SecretProofTransactionBuilder = SecretProofTransactionBuilder;
