// catbuffer-typescript/builders/TransferTransactionBuilder.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var GeneratorUtils_1 = require("./GeneratorUtils");
var TransactionBuilder_1 = require("./TransactionBuilder");
var TransferTransactionBodyBuilder_1 = require("./TransferTransactionBodyBuilder");
var TransferTransactionBuilder = (function (_super) {
    __extends(TransferTransactionBuilder, _super);
    function TransferTransactionBuilder(signature, signerPublicKey, version, network, type, fee, deadline, recipientAddress, mosaics, message) {
        var _this = _super.call(this, signature, signerPublicKey, version, network, type, fee, deadline) || this;
        _this.transferTransactionBody = new TransferTransactionBodyBuilder_1.TransferTransactionBodyBuilder(recipientAddress, mosaics, message);
        return _this;
    }
    TransferTransactionBuilder.loadFromBinary = function (payload) {
        var byteArray = Array.from(payload);
        var superObject = TransactionBuilder_1.TransactionBuilder.loadFromBinary(Uint8Array.from(byteArray));
        byteArray.splice(0, superObject.getSize());
        var transferTransactionBody = TransferTransactionBodyBuilder_1.TransferTransactionBodyBuilder.loadFromBinary(Uint8Array.from(byteArray));
        byteArray.splice(0, transferTransactionBody.getSize());
        return new TransferTransactionBuilder(superObject.signature, superObject.signerPublicKey, superObject.version, superObject.network, superObject.type, superObject.fee, superObject.deadline, transferTransactionBody.recipientAddress, transferTransactionBody.mosaics, transferTransactionBody.message);
    };
    TransferTransactionBuilder.prototype.getRecipientAddress = function () {
        return this.transferTransactionBody.getRecipientAddress();
    };
    TransferTransactionBuilder.prototype.getTransferTransactionBody_Reserved1 = function () {
        return this.transferTransactionBody.getTransferTransactionBody_Reserved1();
    };
    TransferTransactionBuilder.prototype.getMosaics = function () {
        return this.transferTransactionBody.getMosaics();
    };
    TransferTransactionBuilder.prototype.getMessage = function () {
        return this.transferTransactionBody.getMessage();
    };
    TransferTransactionBuilder.prototype.getSize = function () {
        var size = _super.prototype.getSize.call(this);
        size += this.transferTransactionBody.getSize();
        return size;
    };
    TransferTransactionBuilder.prototype.serialize = function () {
        var newArray = Uint8Array.from([]);
        var superBytes = _super.prototype.serialize.call(this);
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, superBytes);
        var transferTransactionBodyBytes = this.transferTransactionBody.serialize();
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, transferTransactionBodyBytes);
        return newArray;
    };
    return TransferTransactionBuilder;
}(TransactionBuilder_1.TransactionBuilder));
exports.TransferTransactionBuilder = TransferTransactionBuilder;
