// symbol-sdk/dist/src/infrastructure/MosaicHttp.js
"use strict";
/*
 * Copyright 2019 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const operators_1 = require("rxjs/operators");
const symbol_openapi_typescript_node_client_1 = require("symbol-openapi-typescript-node-client");
const PublicAccount_1 = require("../model/account/PublicAccount");
const MosaicFlags_1 = require("../model/mosaic/MosaicFlags");
const MosaicId_1 = require("../model/mosaic/MosaicId");
const MosaicInfo_1 = require("../model/mosaic/MosaicInfo");
const UInt64_1 = require("../model/UInt64");
const Http_1 = require("./Http");
/**
 * Mosaic http repository.
 *
 * @since 1.0
 */
class MosaicHttp extends Http_1.Http {
    /**
     * Constructor
     * @param url
     * @param networkType
     */
    constructor(url, networkType) {
        super(url);
        this.mosaicRoutesApi = new symbol_openapi_typescript_node_client_1.MosaicRoutesApi(url);
        this.networkTypeObservable = this.createNetworkTypeObservable(networkType);
        this.mosaicRoutesApi.useQuerystring = true;
    }
    /**
     * Gets the MosaicInfo for a given mosaicId
     * @param mosaicId - Mosaic id
     * @returns Observable<MosaicInfo>
     */
    getMosaic(mosaicId) {
        return this.networkTypeObservable.pipe(operators_1.mergeMap((networkType) => this.call(this.mosaicRoutesApi.getMosaic(mosaicId.toHex()), (body) => this.toMosaicInfo(body, networkType))));
    }
    /**
     * Gets MosaicInfo for different mosaicIds.
     * @param mosaicIds - Array of mosaic ids
     * @returns Observable<MosaicInfo[]>
     */
    getMosaics(mosaicIds) {
        const ids = new symbol_openapi_typescript_node_client_1.MosaicIds();
        ids.mosaicIds = mosaicIds.map((id) => id.toHex());
        return this.networkTypeObservable.pipe(operators_1.mergeMap((networkType) => this.call(this.mosaicRoutesApi.getMosaics(ids), (body) => body.map((b) => this.toMosaicInfo(b, networkType)))));
    }
    /**
     * Gets an array of mosaics created for a given account address.
     * @summary Get mosaics created by an account
     * @param address Account address.
     */
    getMosaicsFromAccount(address) {
        return this.networkTypeObservable.pipe(operators_1.mergeMap((networkType) => this.call(this.mosaicRoutesApi.getMosaicsFromAccount(address.plain()), (body) => body.mosaics.map((b) => this.toMosaicInfoFromMosaicDto(b, networkType)))));
    }
    /**
     * Gets mosaics created for a given array of addresses.
     * @summary Get mosaics created for given array of addresses
     * @param addresses Array of addresses
     */
    getMosaicsFromAccounts(addresses) {
        const accountIds = new symbol_openapi_typescript_node_client_1.AccountIds();
        accountIds.addresses = addresses.map((address) => address.plain());
        return this.networkTypeObservable.pipe(operators_1.mergeMap((networkType) => this.call(this.mosaicRoutesApi.getMosaicsFromAccounts(accountIds), (body) => body.mosaics.map((b) => this.toMosaicInfoFromMosaicDto(b, networkType)))));
    }
    /**
     * Maps MosaicInfoDTO to MosaicInfo
     *
     * @param mosaicInfo the dto object.
     * @returns the model object
     */
    toMosaicInfo(mosaicInfo, networkType) {
        return new MosaicInfo_1.MosaicInfo(new MosaicId_1.MosaicId(mosaicInfo.mosaic.id), UInt64_1.UInt64.fromNumericString(mosaicInfo.mosaic.supply), UInt64_1.UInt64.fromNumericString(mosaicInfo.mosaic.startHeight), PublicAccount_1.PublicAccount.createFromPublicKey(mosaicInfo.mosaic.ownerPublicKey, networkType), mosaicInfo.mosaic.revision, new MosaicFlags_1.MosaicFlags(mosaicInfo.mosaic.flags), mosaicInfo.mosaic.divisibility, UInt64_1.UInt64.fromNumericString(mosaicInfo.mosaic.duration));
    }
    /**
     * Maps MosaicDTO to MosaicInfo
     *
     * @param mosaicInfo the dto object.
     * @returns the model object
     */
    toMosaicInfoFromMosaicDto(mosaicInfo, networkType) {
        return new MosaicInfo_1.MosaicInfo(new MosaicId_1.MosaicId(mosaicInfo.id), UInt64_1.UInt64.fromNumericString(mosaicInfo.supply), UInt64_1.UInt64.fromNumericString(mosaicInfo.startHeight), PublicAccount_1.PublicAccount.createFromPublicKey(mosaicInfo.ownerPublicKey, networkType), mosaicInfo.revision, new MosaicFlags_1.MosaicFlags(mosaicInfo.flags), mosaicInfo.divisibility, UInt64_1.UInt64.fromNumericString(mosaicInfo.duration));
    }
}
exports.MosaicHttp = MosaicHttp;
