// symbol-sdk/dist/src/infrastructure/RepositoryFactoryHttp.js
"use strict";
/*
 * Copyright 2019 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const rxjs_1 = require("rxjs");
const operators_1 = require("rxjs/operators");
const UInt64_1 = require("../model/UInt64");
const AccountHttp_1 = require("./AccountHttp");
const BlockHttp_1 = require("./BlockHttp");
const ChainHttp_1 = require("./ChainHttp");
const Listener_1 = require("./Listener");
const MetadataHttp_1 = require("./MetadataHttp");
const MosaicHttp_1 = require("./MosaicHttp");
const MultisigHttp_1 = require("./MultisigHttp");
const NamespaceHttp_1 = require("./NamespaceHttp");
const NetworkHttp_1 = require("./NetworkHttp");
const NodeHttp_1 = require("./NodeHttp");
const ReceiptHttp_1 = require("./ReceiptHttp");
const RestrictionAccountHttp_1 = require("./RestrictionAccountHttp");
const RestrictionMosaicHttp_1 = require("./RestrictionMosaicHttp");
const TransactionHttp_1 = require("./TransactionHttp");
/**
 * Receipt http repository.
 *
 */
class RepositoryFactoryHttp {
    /**
     * Constructor
     * @param url the server url.
     * @param networkType optional network type if you don't want to load it from the server.
     * @param generationHash optional node generation hash if you don't want to load it from the server.
     */
    constructor(url, networkType, generationHash) {
        this.url = url;
        this.networkType = networkType ? rxjs_1.of(networkType) : this.createNetworkRepository().getNetworkType().pipe(operators_1.shareReplay(1));
        this.generationHash = generationHash
            ? rxjs_1.of(generationHash)
            : this.createBlockRepository()
                .getBlockByHeight(UInt64_1.UInt64.fromUint(1))
                .pipe(operators_1.map((b) => b.generationHash))
                .pipe(operators_1.shareReplay(1));
    }
    createAccountRepository() {
        return new AccountHttp_1.AccountHttp(this.url);
    }
    createBlockRepository() {
        return new BlockHttp_1.BlockHttp(this.url);
    }
    createChainRepository() {
        return new ChainHttp_1.ChainHttp(this.url);
    }
    createMetadataRepository() {
        return new MetadataHttp_1.MetadataHttp(this.url);
    }
    createMosaicRepository() {
        return new MosaicHttp_1.MosaicHttp(this.url, this.networkType);
    }
    createMultisigRepository() {
        return new MultisigHttp_1.MultisigHttp(this.url, this.networkType);
    }
    createNamespaceRepository() {
        return new NamespaceHttp_1.NamespaceHttp(this.url, this.networkType);
    }
    createNetworkRepository() {
        return new NetworkHttp_1.NetworkHttp(this.url);
    }
    createNodeRepository() {
        return new NodeHttp_1.NodeHttp(this.url);
    }
    createReceiptRepository() {
        return new ReceiptHttp_1.ReceiptHttp(this.url, this.networkType);
    }
    createRestrictionAccountRepository() {
        return new RestrictionAccountHttp_1.RestrictionAccountHttp(this.url);
    }
    createRestrictionMosaicRepository() {
        return new RestrictionMosaicHttp_1.RestrictionMosaicHttp(this.url);
    }
    createTransactionRepository() {
        return new TransactionHttp_1.TransactionHttp(this.url);
    }
    getGenerationHash() {
        return this.generationHash;
    }
    getNetworkType() {
        return this.networkType;
    }
    createListener() {
        return new Listener_1.Listener(this.url);
    }
}
exports.RepositoryFactoryHttp = RepositoryFactoryHttp;
