// symbol-openapi-typescript-node-client/model/aggregateNetworkPropertiesDTO.ts
/*
 * Copyright 2019 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Catapult REST Endpoints
 * OpenAPI Specification of catapult-rest 1.0.20.22
 *
 * The version of the OpenAPI document: 0.8.9
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

import { RequestFile } from '../api';

export class AggregateNetworkPropertiesDTO {
    /**
    * Maximum number of transactions per aggregate.
    */
    'maxTransactionsPerAggregate'?: string;
    /**
    * Maximum number of cosignatures per aggregate.
    */
    'maxCosignaturesPerAggregate'?: string;
    /**
    * Set to true if cosignatures must exactly match component signers. Set to false if cosignatures should be validated externally.
    */
    'enableStrictCosignatureCheck'?: boolean;
    /**
    * Set to true if bonded aggregates should be allowed. Set to false if bonded aggregates should be rejected.
    */
    'enableBondedAggregateSupport'?: boolean;
    /**
    * Maximum lifetime a bonded transaction can have before it expires.
    */
    'maxBondedTransactionLifetime'?: string;

    static discriminator: string | undefined = undefined;

    static attributeTypeMap: Array<{name: string, baseName: string, type: string}> = [
        {
            "name": "maxTransactionsPerAggregate",
            "baseName": "maxTransactionsPerAggregate",
            "type": "string"
        },
        {
            "name": "maxCosignaturesPerAggregate",
            "baseName": "maxCosignaturesPerAggregate",
            "type": "string"
        },
        {
            "name": "enableStrictCosignatureCheck",
            "baseName": "enableStrictCosignatureCheck",
            "type": "boolean"
        },
        {
            "name": "enableBondedAggregateSupport",
            "baseName": "enableBondedAggregateSupport",
            "type": "boolean"
        },
        {
            "name": "maxBondedTransactionLifetime",
            "baseName": "maxBondedTransactionLifetime",
            "type": "string"
        }    ];

    static getAttributeTypeMap() {
        return AggregateNetworkPropertiesDTO.attributeTypeMap;
    }
}
