// symbol-openapi-typescript-node-client/model/embeddedMosaicAddressRestrictionTransactionDTO.ts
/*
 * Copyright 2019 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Catapult REST Endpoints
 * OpenAPI Specification of catapult-rest 1.0.20.22
 *
 * The version of the OpenAPI document: 0.8.9
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

import { RequestFile } from '../api';
import { EmbeddedTransactionDTO } from './embeddedTransactionDTO';
import { MosaicAddressRestrictionTransactionBodyDTO } from './mosaicAddressRestrictionTransactionBodyDTO';
import { NetworkTypeEnum } from './networkTypeEnum';

export class EmbeddedMosaicAddressRestrictionTransactionDTO {
    /**
    * Public key.
    */
    'signerPublicKey': string;
    /**
    * Entity version.
    */
    'version': number;
    'network': NetworkTypeEnum;
    'type': number;
    /**
    * Mosaic identifier. If the most significant bit of byte 0 is set, a namespaceId (alias) is used instead of the real mosaic identifier. 
    */
    'mosaicId': string;
    /**
    * Restriction key.
    */
    'restrictionKey': string;
    /**
    * Restriction value.
    */
    'previousRestrictionValue': string;
    /**
    * Restriction value.
    */
    'newRestrictionValue': string;
    /**
    * Address expressed in hexadecimal base. If the bit 0 of byte 0 is not set (like in 0x90), then it is a regular address. Else (e.g. 0x91) it represents a namespace id which starts at byte 1. 
    */
    'targetAddress': string;

    static discriminator: string | undefined = undefined;

    static attributeTypeMap: Array<{name: string, baseName: string, type: string}> = [
        {
            "name": "signerPublicKey",
            "baseName": "signerPublicKey",
            "type": "string"
        },
        {
            "name": "version",
            "baseName": "version",
            "type": "number"
        },
        {
            "name": "network",
            "baseName": "network",
            "type": "NetworkTypeEnum"
        },
        {
            "name": "type",
            "baseName": "type",
            "type": "number"
        },
        {
            "name": "mosaicId",
            "baseName": "mosaicId",
            "type": "string"
        },
        {
            "name": "restrictionKey",
            "baseName": "restrictionKey",
            "type": "string"
        },
        {
            "name": "previousRestrictionValue",
            "baseName": "previousRestrictionValue",
            "type": "string"
        },
        {
            "name": "newRestrictionValue",
            "baseName": "newRestrictionValue",
            "type": "string"
        },
        {
            "name": "targetAddress",
            "baseName": "targetAddress",
            "type": "string"
        }    ];

    static getAttributeTypeMap() {
        return EmbeddedMosaicAddressRestrictionTransactionDTO.attributeTypeMap;
    }
}
