// symbol-openapi-typescript-node-client/model/mosaicNetworkPropertiesDTO.ts
/*
 * Copyright 2019 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Catapult REST Endpoints
 * OpenAPI Specification of catapult-rest 1.0.20.22
 *
 * The version of the OpenAPI document: 0.8.9
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

import { RequestFile } from '../api';

export class MosaicNetworkPropertiesDTO {
    /**
    * Maximum number of mosaics that an account can own.
    */
    'maxMosaicsPerAccount'?: string;
    /**
    * Maximum mosaic duration.
    */
    'maxMosaicDuration'?: string;
    /**
    * Maximum mosaic divisibility.
    */
    'maxMosaicDivisibility'?: string;
    /**
    * Public key.
    */
    'mosaicRentalFeeSinkPublicKey'?: string;
    /**
    * Mosaic rental fee.
    */
    'mosaicRentalFee'?: string;

    static discriminator: string | undefined = undefined;

    static attributeTypeMap: Array<{name: string, baseName: string, type: string}> = [
        {
            "name": "maxMosaicsPerAccount",
            "baseName": "maxMosaicsPerAccount",
            "type": "string"
        },
        {
            "name": "maxMosaicDuration",
            "baseName": "maxMosaicDuration",
            "type": "string"
        },
        {
            "name": "maxMosaicDivisibility",
            "baseName": "maxMosaicDivisibility",
            "type": "string"
        },
        {
            "name": "mosaicRentalFeeSinkPublicKey",
            "baseName": "mosaicRentalFeeSinkPublicKey",
            "type": "string"
        },
        {
            "name": "mosaicRentalFee",
            "baseName": "mosaicRentalFee",
            "type": "string"
        }    ];

    static getAttributeTypeMap() {
        return MosaicNetworkPropertiesDTO.attributeTypeMap;
    }
}
