// symbol-openapi-typescript-node-client/model/nodeInfoDTO.ts
/*
 * Copyright 2019 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Catapult REST Endpoints
 * OpenAPI Specification of catapult-rest 1.0.20.22
 *
 * The version of the OpenAPI document: 0.8.9
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

import { RequestFile } from '../api';
import { RolesTypeEnum } from './rolesTypeEnum';

export class NodeInfoDTO {
    /**
    * Version of the application.
    */
    'version': number;
    /**
    * Public key.
    */
    'publicKey': string;
    'networkGenerationHash': string;
    'roles': RolesTypeEnum;
    /**
    * Port used for the communication.
    */
    'port': number;
    'networkIdentifier': number;
    /**
    * Node friendly name.
    */
    'friendlyName': string;
    /**
    * Node IP address.
    */
    'host': string;

    static discriminator: string | undefined = undefined;

    static attributeTypeMap: Array<{name: string, baseName: string, type: string}> = [
        {
            "name": "version",
            "baseName": "version",
            "type": "number"
        },
        {
            "name": "publicKey",
            "baseName": "publicKey",
            "type": "string"
        },
        {
            "name": "networkGenerationHash",
            "baseName": "networkGenerationHash",
            "type": "string"
        },
        {
            "name": "roles",
            "baseName": "roles",
            "type": "RolesTypeEnum"
        },
        {
            "name": "port",
            "baseName": "port",
            "type": "number"
        },
        {
            "name": "networkIdentifier",
            "baseName": "networkIdentifier",
            "type": "number"
        },
        {
            "name": "friendlyName",
            "baseName": "friendlyName",
            "type": "string"
        },
        {
            "name": "host",
            "baseName": "host",
            "type": "string"
        }    ];

    static getAttributeTypeMap() {
        return NodeInfoDTO.attributeTypeMap;
    }
}
