// symbol-openapi-typescript-node-client/model/pluginsPropertiesDTO.ts
/*
 * Copyright 2019 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Catapult REST Endpoints
 * OpenAPI Specification of catapult-rest 1.0.20.22
 *
 * The version of the OpenAPI document: 0.8.9
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

import { RequestFile } from '../api';
import { AccountLinkNetworkPropertiesDTO } from './accountLinkNetworkPropertiesDTO';
import { AccountRestrictionNetworkPropertiesDTO } from './accountRestrictionNetworkPropertiesDTO';
import { AggregateNetworkPropertiesDTO } from './aggregateNetworkPropertiesDTO';
import { HashLockNetworkPropertiesDTO } from './hashLockNetworkPropertiesDTO';
import { MetadataNetworkPropertiesDTO } from './metadataNetworkPropertiesDTO';
import { MosaicNetworkPropertiesDTO } from './mosaicNetworkPropertiesDTO';
import { MosaicRestrictionNetworkPropertiesDTO } from './mosaicRestrictionNetworkPropertiesDTO';
import { MultisigNetworkPropertiesDTO } from './multisigNetworkPropertiesDTO';
import { NamespaceNetworkPropertiesDTO } from './namespaceNetworkPropertiesDTO';
import { SecretLockNetworkPropertiesDTO } from './secretLockNetworkPropertiesDTO';
import { TransferNetworkPropertiesDTO } from './transferNetworkPropertiesDTO';

/**
* Plugin related configuration properties.
*/
export class PluginsPropertiesDTO {
    'accountlink'?: AccountLinkNetworkPropertiesDTO;
    'aggregate'?: AggregateNetworkPropertiesDTO;
    'lockhash'?: HashLockNetworkPropertiesDTO;
    'locksecret'?: SecretLockNetworkPropertiesDTO;
    'metadata'?: MetadataNetworkPropertiesDTO;
    'mosaic'?: MosaicNetworkPropertiesDTO;
    'multisig'?: MultisigNetworkPropertiesDTO;
    'namespace'?: NamespaceNetworkPropertiesDTO;
    'restrictionaccount'?: AccountRestrictionNetworkPropertiesDTO;
    'restrictionmosaic'?: MosaicRestrictionNetworkPropertiesDTO;
    'transfer'?: TransferNetworkPropertiesDTO;

    static discriminator: string | undefined = undefined;

    static attributeTypeMap: Array<{name: string, baseName: string, type: string}> = [
        {
            "name": "accountlink",
            "baseName": "accountlink",
            "type": "AccountLinkNetworkPropertiesDTO"
        },
        {
            "name": "aggregate",
            "baseName": "aggregate",
            "type": "AggregateNetworkPropertiesDTO"
        },
        {
            "name": "lockhash",
            "baseName": "lockhash",
            "type": "HashLockNetworkPropertiesDTO"
        },
        {
            "name": "locksecret",
            "baseName": "locksecret",
            "type": "SecretLockNetworkPropertiesDTO"
        },
        {
            "name": "metadata",
            "baseName": "metadata",
            "type": "MetadataNetworkPropertiesDTO"
        },
        {
            "name": "mosaic",
            "baseName": "mosaic",
            "type": "MosaicNetworkPropertiesDTO"
        },
        {
            "name": "multisig",
            "baseName": "multisig",
            "type": "MultisigNetworkPropertiesDTO"
        },
        {
            "name": "namespace",
            "baseName": "namespace",
            "type": "NamespaceNetworkPropertiesDTO"
        },
        {
            "name": "restrictionaccount",
            "baseName": "restrictionaccount",
            "type": "AccountRestrictionNetworkPropertiesDTO"
        },
        {
            "name": "restrictionmosaic",
            "baseName": "restrictionmosaic",
            "type": "MosaicRestrictionNetworkPropertiesDTO"
        },
        {
            "name": "transfer",
            "baseName": "transfer",
            "type": "TransferNetworkPropertiesDTO"
        }    ];

    static getAttributeTypeMap() {
        return PluginsPropertiesDTO.attributeTypeMap;
    }
}
